## The Client

XYZ is a private firm in US. Due to remarkable growth in the Cab Industry in last few years and multiple key players in the market, it is planning for an investment in Cab industry and as per their Go-to-Market(G2M) strategy they want to understand the market before taking final decision.

Project delivery:

You have been provided with multiple data sets that contains information on 2 cab companies. Each file (data set) provided represents different aspects of the customer profile. XYZ is interested in using your actionable insights to help them identify the right company to make their investment.

The outcome of your delivery will be a presentation to XYZ’s Executive team. This presentation will be judged based on the visuals provided, the quality of your analysis and the value of your recommendations and insights. 

## Data Set:

You have been provided 4 individual data sets. Time period of data is from 31/01/2016 to 31/12/2018.

Below are the list of datasets which are provided for the analysis:

Cab_Data.csv – this file includes details of transaction for 2 cab companies

Customer_ID.csv – this is a mapping table that contains a unique identifier which links the customer’s demographic details

Transaction_ID.csv – this is a mapping table that contains transaction to customer mapping and payment mode

City.csv – this file contains list of US cities, their population and number of cab users

You should fully investigate and understand each data set.

Review the Source Documentation
Understand the field names and data types
Identify relationships across the files
Field/feature transformations
Determine which files should be joined versus which ones should be appended
Create master data and explain the relationship
Identify and remove duplicates
Perform other analysis like NA value and outlier detection
 

Whatever and how many slides you prepare(Be creative and come up with meaningful insight):

The idea is to create a hypothesis, engage with the data, think critically, and use various analytical approaches to produce unique insights.

You are not limited to only utilizing the data you have been provided.

We encourage you to find third party data sets which correspond to the overall theme and geographical properties of the data provided.  For Example: you can leverage US holiday data/ weather data

Also, do research on overall cab industry in US and try to relate that with the trend in data

Analysis

Create multiple hypothesis and investigate:

You will need to generate 5-7 hypothesis initially to investigate as some will not prove what you are expecting.

In [1]:
# libraries

import pandas as pd
import numpy as np

In [ ]:
# import datasets

cab_df = pd.read_csv("Cab_Data.csv")
city_df = pd.read_csv("City.csv")
customer_df = pd.read_csv("Customer_ID.csv")
transaction_df = pd.read_csv("Transaction_ID.csv")

In [5]:
cab_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 359392 entries, 0 to 359391
Data columns (total 7 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Transaction ID  359392 non-null  int64  
 1   Date of Travel  359392 non-null  int64  
 2   Company         359392 non-null  object 
 3   City            359392 non-null  object 
 4   KM Travelled    359392 non-null  float64
 5   Price Charged   359392 non-null  float64
 6   Cost of Trip    359392 non-null  float64
dtypes: float64(3), int64(2), object(2)
memory usage: 19.2+ MB


In [6]:
city_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   City        20 non-null     object
 1   Population  20 non-null     object
 2   Users       20 non-null     object
dtypes: object(3)
memory usage: 612.0+ bytes


In [7]:
customer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49171 entries, 0 to 49170
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Customer ID         49171 non-null  int64 
 1   Gender              49171 non-null  object
 2   Age                 49171 non-null  int64 
 3   Income (USD/Month)  49171 non-null  int64 
dtypes: int64(3), object(1)
memory usage: 1.5+ MB


In [8]:
transaction_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 440098 entries, 0 to 440097
Data columns (total 3 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   Transaction ID  440098 non-null  int64 
 1   Customer ID     440098 non-null  int64 
 2   Payment_Mode    440098 non-null  object
dtypes: int64(2), object(1)
memory usage: 10.1+ MB


All four files do not contain any missing values.  We may need to make some adjustments to the data types
Cab_df - transaction id (object) 
city_df - population and users (numeric int64)
customer_df - Customer ID (object or string)
transaction_df - transaction ID and customer ID both object or string

The formatting for Date of Travel is not correct so I left it as is.  I did not undersand what "42377" represented. I guess it does not matter since it's just a point in time.

After these changes, let's look at the first few rows of each file. Verify that data types are correct now.

In [ ]:
cab_df["Transaction ID"] = cab_df["Transaction ID"].astype(str)
city_df["Population"] = city_df["Population"].str.replace(",", "").astype(int)   #series
city_df["Users"] = city_df["Users"].str.replace(",", "").astype(int)
customer_df["Customer ID"] = customer_df["Customer ID"].astype(str)
transaction_df["Customer ID"] = transaction_df["Customer ID"].astype(str)
transaction_df["Transaction ID"] = transaction_df["Transaction ID"].astype(str)

In [24]:
cab_df.head()

Transaction ID  Date of Travel   Company        City  KM Travelled  \
0       10000011           42377  Pink Cab  ATLANTA GA         30.45   
1       10000012           42375  Pink Cab  ATLANTA GA         28.62   
2       10000013           42371  Pink Cab  ATLANTA GA          9.04   
3       10000014           42376  Pink Cab  ATLANTA GA         33.17   
4       10000015           42372  Pink Cab  ATLANTA GA          8.73   

   Price Charged  Cost of Trip  
0         370.95       313.635  
1         358.52       334.854  
2         125.20        97.632  
3         377.40       351.602  
4         114.62        97.776

In [25]:
city_df.head()

City  Population   Users
0     NEW YORK NY     8405837  302149
1      CHICAGO IL     1955130  164468
2  LOS ANGELES CA     1595037  144132
3        MIAMI FL     1339155   17675
4  SILICON VALLEY     1177609   27247

In [26]:
customer_df.head()

Customer ID Gender  Age  Income (USD/Month)
0       29290   Male   28               10813
1       27703   Male   27                9237
2       28712   Male   53               11242
3       28020   Male   23               23327
4       27182   Male   33                8536

In [27]:
transaction_df.head()

Transaction ID Customer ID Payment_Mode
0       10000011       29290         Card
1       10000012       27703         Card
2       10000013       28712         Cash
3       10000014       28020         Cash
4       10000015       27182         Card

## Understand the field names and data types
EDA - Begin with using describe() method
Identify relationships across the files
Field/feature transformations

In [28]:
cab_df.describe()

Date of Travel   KM Travelled  Price Charged   Cost of Trip
count   359392.000000  359392.000000  359392.000000  359392.000000
mean     42964.067998      22.567254     423.443311     286.190113
std        307.467197      12.233526     274.378911     157.993661
min      42371.000000       1.900000      15.600000      19.000000
25%      42697.000000      12.000000     206.437500     151.200000
50%      42988.000000      22.440000     386.360000     282.480000
75%      43232.000000      32.960000     583.660000     413.683200
max      43465.000000      48.000000    2048.030000     691.200000

The average distance travelled per cab was 22.6 KM with an average fare of $423.44.

In [29]:
city_df.describe()

Population          Users
count  2.000000e+01      20.000000
mean   1.231592e+06   64520.650000
std    1.740127e+06   83499.375289
min    2.489680e+05    3643.000000
25%    6.086372e+05   11633.250000
50%    7.845590e+05   23429.000000
75%    1.067041e+06   91766.000000
max    8.405837e+06  302149.000000

In the selected 20 cities, 64521 users are taking a cab.

In [30]:
customer_df.describe()

Age  Income (USD/Month)
count  49171.000000        49171.000000
mean      35.363121        15015.631856
std       12.599066         8002.208253
min       18.000000         2000.000000
25%       25.000000         8289.500000
50%       33.000000        14656.000000
75%       42.000000        21035.000000
max       65.000000        35000.000000

The average age of the user (of cabs) is 35 years old with an average income of $15,015.63.  The youngest is 18 years old.  The oldest is 65 years old.

In [31]:
transaction_df.describe()

Transaction ID Customer ID Payment_Mode
count          440098      440098       440098
unique         440098       49171            2
top          10000011         494         Card
freq                1          54       263991

There are 440,098 unique transactaions.  There are users taking a cab more than once.

### Determine which files should be joined versus which ones should be appended
Create master data and explain the relationship

The relationship starts with `city` field in `cab_df`. It will be joined (one to one) to the `city` field in `city_df`. Next, the `transaction id` field in `transaction_df` is joined (one to one) to `transaction_id` in `city_df`. Finally, `customer_id` in `transaction_df` is joined to `customer_id` in `customer_df`

In [32]:
step1 = pd.merge(cab_df, city_df, how = "left", on =["City", "City"])
step1

Transaction ID  Date of Travel     Company           City  \
0            10000011           42377    Pink Cab     ATLANTA GA   
1            10000012           42375    Pink Cab     ATLANTA GA   
2            10000013           42371    Pink Cab     ATLANTA GA   
3            10000014           42376    Pink Cab     ATLANTA GA   
4            10000015           42372    Pink Cab     ATLANTA GA   
...               ...             ...         ...            ...   
359387       10440101           43108  Yellow Cab  WASHINGTON DC   
359388       10440104           43104  Yellow Cab  WASHINGTON DC   
359389       10440105           43105  Yellow Cab  WASHINGTON DC   
359390       10440106           43105  Yellow Cab  WASHINGTON DC   
359391       10440107           43102  Yellow Cab  WASHINGTON DC   

        KM Travelled  Price Charged  Cost of Trip  Population   Users  
0              30.45         370.95      313.6350      814885   24701  
1              28.62         358.52      334.8540      814885   24701  
2               9.04         125.20       97.6320      814885   24701  
3              33.17         377.40      351.6020      814885   24701  
4               8.73         114.62       97.7760      814885   24701  
...              ...            ...           ...         ...     ...  
359387          4.80          69.24       63.3600      418859  127001  
359388          8.40         113.75      106.8480      418859  127001  
359389         27.75         437.07      349.6500      418859  127001  
359390          8.80         146.19      114.0480      418859  127001  
359391         12.76         191.58      177.6192      418859  127001  

[359392 rows x 9 columns]

In [34]:
step2 = pd.merge(step1, transaction_df, how ="left", on= ["Transaction ID", "Transaction ID"])
step2

Transaction ID  Date of Travel     Company           City  \
0            10000011           42377    Pink Cab     ATLANTA GA   
1            10000012           42375    Pink Cab     ATLANTA GA   
2            10000013           42371    Pink Cab     ATLANTA GA   
3            10000014           42376    Pink Cab     ATLANTA GA   
4            10000015           42372    Pink Cab     ATLANTA GA   
...               ...             ...         ...            ...   
359387       10440101           43108  Yellow Cab  WASHINGTON DC   
359388       10440104           43104  Yellow Cab  WASHINGTON DC   
359389       10440105           43105  Yellow Cab  WASHINGTON DC   
359390       10440106           43105  Yellow Cab  WASHINGTON DC   
359391       10440107           43102  Yellow Cab  WASHINGTON DC   

        KM Travelled  Price Charged  Cost of Trip  Population   Users  \
0              30.45         370.95      313.6350      814885   24701   
1              28.62         358.52      334.8540      814885   24701   
2               9.04         125.20       97.6320      814885   24701   
3              33.17         377.40      351.6020      814885   24701   
4               8.73         114.62       97.7760      814885   24701   
...              ...            ...           ...         ...     ...   
359387          4.80          69.24       63.3600      418859  127001   
359388          8.40         113.75      106.8480      418859  127001   
359389         27.75         437.07      349.6500      418859  127001   
359390          8.80         146.19      114.0480      418859  127001   
359391         12.76         191.58      177.6192      418859  127001   

       Customer ID Payment_Mode  
0            29290         Card  
1            27703         Card  
2            28712         Cash  
3            28020         Cash  
4            27182         Card  
...            ...          ...  
359387       52392         Cash  
359388       53286         Cash  
359389       52265         Cash  
359390       52175         Card  
359391       52917         Card  

[359392 rows x 11 columns]

In [37]:
result_df = pd.merge(step2, customer_df, how = "left", on= ["Customer ID", "Customer ID"])
result_df

Transaction ID  Date of Travel     Company           City  \
0            10000011           42377    Pink Cab     ATLANTA GA   
1            10000012           42375    Pink Cab     ATLANTA GA   
2            10000013           42371    Pink Cab     ATLANTA GA   
3            10000014           42376    Pink Cab     ATLANTA GA   
4            10000015           42372    Pink Cab     ATLANTA GA   
...               ...             ...         ...            ...   
359387       10440101           43108  Yellow Cab  WASHINGTON DC   
359388       10440104           43104  Yellow Cab  WASHINGTON DC   
359389       10440105           43105  Yellow Cab  WASHINGTON DC   
359390       10440106           43105  Yellow Cab  WASHINGTON DC   
359391       10440107           43102  Yellow Cab  WASHINGTON DC   

        KM Travelled  Price Charged  Cost of Trip  Population   Users  \
0              30.45         370.95      313.6350      814885   24701   
1              28.62         358.52      334.8540      814885   24701   
2               9.04         125.20       97.6320      814885   24701   
3              33.17         377.40      351.6020      814885   24701   
4               8.73         114.62       97.7760      814885   24701   
...              ...            ...           ...         ...     ...   
359387          4.80          69.24       63.3600      418859  127001   
359388          8.40         113.75      106.8480      418859  127001   
359389         27.75         437.07      349.6500      418859  127001   
359390          8.80         146.19      114.0480      418859  127001   
359391         12.76         191.58      177.6192      418859  127001   

       Customer ID Payment_Mode Gender  Age  Income (USD/Month)  
0            29290         Card   Male   28               10813  
1            27703         Card   Male   27                9237  
2            28712         Cash   Male   53               11242  
3            28020         Cash   Male   23               23327  
4            27182         Card   Male   33                8536  
...            ...          ...    ...  ...                 ...  
359387       52392         Cash   Male   24               15651  
359388       53286         Cash   Male   32                6528  
359389       52265         Cash   Male   56                7966  
359390       52175         Card   Male   32                6423  
359391       52917         Card   Male   20               11284  

[359392 rows x 14 columns]

In [ ]:
# double check that joins did not create null values
result_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 359392 entries, 0 to 359391
Data columns (total 14 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Transaction ID      359392 non-null  object 
 1   Date of Travel      359392 non-null  int64  
 2   Company             359392 non-null  object 
 3   City                359392 non-null  object 
 4   KM Travelled        359392 non-null  float64
 5   Price Charged       359392 non-null  float64
 6   Cost of Trip        359392 non-null  float64
 7   Population          359392 non-null  int32  
 8   Users               359392 non-null  int32  
 9   Customer ID         359392 non-null  object 
 10  Payment_Mode        359392 non-null  object 
 11  Gender              359392 non-null  object 
 12  Age                 359392 non-null  int64  
 13  Income (USD/Month)  359392 non-null  int64  
dtypes: float64(3), int32(2), int64(3), object(6)
memory usage: 35.6+ MB


In [ ]:
# check for duplicate observations. if so, then use df.drop_duplicates()

duplicates = result_df[result_df.duplicated()]
duplicates

Empty DataFrame
Columns: [Transaction ID, Date of Travel, Company, City, KM Travelled, Price Charged, Cost of Trip, Population, Users, Customer ID, Payment_Mode, Gender, Age, Income (USD/Month)]
Index: []

In [ ]:
# check for NA values

result_df.isna().sum()

Transaction ID        0
Date of Travel        0
Company               0
City                  0
KM Travelled          0
Price Charged         0
Cost of Trip          0
Population            0
Users                 0
Customer ID           0
Payment_Mode          0
Gender                0
Age                   0
Income (USD/Month)    0
dtype: int64